### ISSUE TITLE
Show sample citations & bib entries for PRs that validate

### ISSUE THREAD

**Issue by adam3smith** [citation-style-language/Sheldon#13]
Just going to do a simple mock-up:
Currently Sheldon posts this when a PR gets opened:

> Awesome! You just created a pull request to the Citation Styles Language styles repository. One of our human volunteers will try to get in touch soon (usually within a week). In the meantime, I will run some automated checks. You should be notified of the results in a few minutes.
> 
> If you haven't done so yet, please make sure your style [validates](https://github.com/citation-style-language/styles/wiki/Validation) and follows all our other [Style Requirements](https://github.com/citation-style-language/styles/wiki/Style-Requirements).
> 
> To update this pull request, visit the "Files changed" tab above, and click on the pencil icon (see below) in the top-right corner of your style to start editing.
> 
> <img width="274" src="https://user-images.githubusercontent.com/77951/34369455-bd50432c-eab3-11e7-945e-6a6147dfc507.png">
> 
> If you have any questions, please leave a comment and we'll get back to you. While we usually respond in English, feel free to write in whatever language you're most comfortable.

Followed by this when the PR passes our tests:

> :smiley: Your submission passed all our automated tests.

I'd suggest leaving the first message as is, but would like the second one to be something like:

> :smiley: Your submission passed all our automated tests.
> 
> Your style will produce the following output:
> **Citations:** (Campbell & Pedersen, 2007; Mares, 2001)
> 
> **Bibliography:** 
> Campbell, J. L., & Pedersen, O. K. (2007). The varieties of capitalism and hybrid success. *Comparative Political Studies*, 40(3), 307–332. https://doi.org/10.1177/0010414006286542
> Mares, I. (2001). Firms and the welfare state: When, why, and how does social policy matter to employers? In P. A. Hall & D. Soskice (Eds.), *Varieties of capitalism. The institutional foundations of comparative advantage* (pp. 184–213). New York: Oxford University Press.

Where citations and bibliography are generated from the new style (I was thinking we could have the data in a public Zotero group and simply use API calls to Zotero, but agnostic about the method to be used.

**Comment by rmzelle** [comment]
(and, as a second optional layer on top, a diff versus an appropriate comparison style could also be very useful, e.g. using something like https://github.com/kpdecker/jsdiff, which can generate output such as

![image](https://user-images.githubusercontent.com/77951/56819781-7270a700-6818-11e9-9e81-b16a9332797e.png)
)

**Comment by retorquere** [comment]
What would be the benefit to pulling the data from a public group vs having it in the text fixtures?

**Comment by adam3smith** [comment]
This was purely pragmatic -- having them in the fixture is fine too. I was thinking of going through the Zotero API as one option, and for that purpose having them items available in a group makes getting the bib possible with a single `GET` call.

**Comment by retorquere** [comment]
Alright, no major difference between the two anyhow. So, the main ask here is to

1. Render a citation/bibliography with the PR
2. Render a citation/bibliography from master
3. Diff them

(with the rendering done using citeproc-js I assume)

?

**Comment by adam3smith** [comment]
Yes, I think that's basically it. I think we'll only want to do this when the other tests pass, though.

**Comment by retorquere** [comment]
Right. I'm making some headway, but citeproc is a little sparsely documented, so I'm going through their test cases to piece things together.

**Comment by retorquere** [comment]
What is the relation between the spec tests in the `styles` repo and the `test-suite` repo?

**Comment by retorquere** [comment]
Alright, I have diffs rendering, but citeproc in its various incarnations could really do with better documentation. Oy vey. I've stuck to citeproc-ruby because the output really shouldn't differ from citeproc-js and the existing tests used Ruby.

**Comment by adam3smith** [comment]
Sorry, hadn't seen the question above: the test-suite tests correct citeproc behavior and has no relationship to the styles repo. Every test includes its own style.

**Comment by retorquere** [comment]
Any thoughts on what to use as the items to render?

**Comment by adam3smith** [comment]
Here's a set of four that I like (recycled from the CSL editor) https://gist.github.com/adam3smith/7f0c65f116d5e17df0c23901198f42c7

**Comment by retorquere** [comment]
It's been a while since I used Ruby -- how are `describe` et al brought into the current namespace in `styles_spec.rb`? I intended to use an `after(:all)` hook, but rspec complains it can't find `after`.

**Comment by retorquere** [comment]
Ah never mind, I can just use travis hooks.

**Comment by retorquere** [comment]
I'm still not happy with the way things look when added as a comment. I've put up a sample [here](https://github.com/retorquere/styles/pull/1) (apologies for the junk PR I accidentally opened on the actual styles repo)

**Comment by adam3smith** [comment]
What are you not happy with? You could maybe quote it as in the examples above, but I think functionally this is very nice already

**Comment by retorquere** [comment]
I've added a 2nd sample that quotes it. It still looks a little crowded to me, but if it's good enough for you guys, that's what counts.

I'm still playing with diffy to output decent looking diffs, but GH comments are pretty narrow and it gets unreadable fast. Differ does inline colored diffs which are easier to read, but the library hasn't seen updates in 8 years, and issues on the repo don't get picked up. I'm hesitant to bake in a tech debt, OTOH, it does work.

**Comment by retorquere** [comment]
Can anyone point me to a style that changed in a way that the sample references might show a difference in rendering? The random samples I looked at did stuff like et-al fixes that wouldn't trigger on the number of authors in the samples, or for other reasons showed no differences.

**Comment by adam3smith** [comment]
Here are two that would show up in diffs, though I think only for the journal article in both cases:
https://github.com/citation-style-language/styles/pull/4037
https://github.com/citation-style-language/styles/pull/4064

**Comment by adam3smith** [comment]
I like the quoted look -- once we add some surrounding text, I think that'll be great.

**Comment by retorquere** [comment]
Alright - you could start thinking about what you want to show when the render differs and when not.

I could perhaps have the rspec tests only test the changed files in a PR, but there are two sets of styles being tested, and I don't understand the difference between them.

**Comment by retorquere** [comment]
Alright, I've added a new sample with 3 changes; one where a style changed but the output did not, one where the style changed and the output also changed, and one where a new style was added.

Old or not, differ is the only one that I could find that did char-by-char coloration of diffs. Thoughts? Also thoughts on what the diff should diff? Text diffs are going to be more readable than markup diffs, but then markup changes would be lost in the report.

**Comment by retorquere** [comment]
I mean I see tests on "dependents" and "independents" but I don't know what those terms mean in this context.

**Comment by rmzelle** [comment]
> there are two sets of styles being tested, and I don't understand the difference between them.
> ...
> I mean I see tests on "dependents" and "independents" but I don't know what those terms mean in this context.

Are you familiar with dependent CSL styles at all (like https://github.com/citation-style-language/styles/blob/master/dependent/nature-biotechnology.csl)? Dependent CSL styles inherit their style format from the referenced "independent-parent", and are heavily used for large publishers like Elsevier and Springer Nature that have thousands of journals that only use a handful of distinct citation formats. They're stored in the "dependent" directory of the "styles" repo. See also https://docs.citationstyles.org/en/stable/specification.html#file-types.

> differ is the only one that I could find that did char-by-char coloration of diffs.

jsdiff does too (although it currently can't diff markup differences). See http://incaseofstairs.com/jsdiff/.

**Comment by rmzelle** [comment]
> It still looks a little crowded to me

Especially if we envision rendering citations for every commit in a pull request, we might also want to collapse things a bit, e.g.:

<details><summary><b>Changed: apa.csl</b></summary>
<p>
(“CSL search by example,” 2012; Hancké, Rhodes, & Thatcher, 2007)<br>
(Fenner et al., 2019; Mares, 2001)

CSL search by example. (2012). Retrieved December 15, 2012, from Citation Style Editor website: http://editor.citationstyles.org/searchByExample/
Fenner, M., Crosas, M., Grethe, J. S., Kennedy, D., Hermjakob, H., Rocca-Serra, P., … Clark, T. (2019). A data citation roadmap for scholarly data repositories. <i>Scientific Data</i>, <i>6</i>(1), 28. https://doi.org/10.1038/s41597-019-0031-8
Hancké, B., Rhodes, M., & Thatcher, M. (Eds.). (2007). <i>Beyond varieties of capitalism : Conflict, contradiction, and complementarities in the European economy</i>. Oxford and New York: Oxford University Press.
Mares, I. (2001). Firms and the welfare state: When, why, and how does social policy matter to employers? In P. A. Hall & D. Soskice (Eds.), <i>Varieties of capitalism. The institutional foundations of comparative advantage</i> (pp. 184–213). New York: Oxford University Press.
<p>
</details>

(using `<details/>` per https://gist.github.com/joyrexus/16041f2426450e73f5df9391f7f7ae5f and https://github.com/dear-github/dear-github/issues/166)

**Comment by retorquere** [comment]
> Are you familiar with dependent CSL styles at all?

Nope.

> jsdiff does too (although it currently can't diff markup differences). See http://incaseofstairs.com/jsdiff/.

Turns out [GH comments don't allow coloration](https://github.com/github/markup/issues/369). It's possible to do line-by-line coloration by using the diff format, but from my experience that's only of limited use.

**Comment by retorquere** [comment]
I've put up two new samples, one which does line-by-line diffs of html, the other of markdownified html.

**Comment by retorquere** [comment]
Still working on getting the rspec failure output

**Comment by rmzelle** [comment]
I like the line-by-line diffs of HTML best. HTML tags are easier to spot than markdown markup.

Maybe you could tweak the summary labels a little as well? E.g.:

> <details><summary>apa.csl (modified style; unchanged output for sample items)</summary></details>
> <details><summary>international-journal-of-climatology.csl (modified style; changed output)</summary></details>
> <details><summary>junk.csl (new style)</summary></details>

(@adam3smith, let me know if you think that's an improvement)

**Comment by rmzelle** [comment]
And did you already put a limit on the number of styles you're rendering? It's not uncommon to have the occasional pull request that touches several hundred or even thousands of styles, so limiting the render to e.g. no more than 10 styles would be a good idea.

And assuming you'll be handling dependent CSL styles as well, it might be good to:

1. if you reach the limit of styles to render (per above), give preference to independent CSL styles over dependent styles for the styles you do render.
2. We also sometimes replace an independent style by a dependent one, and vice versa. So it would be good if e.g. a replacement of https://github.com/citation-style-language/styles/blob/master/dependent/nature-biotechnology.csl by https://github.com/citation-style-language/styles/blob/master/nature-biotechnology.csl would be recognized as a modified style (the style ID and file name would stay the same in these cases).

**Comment by retorquere** [comment]
I've put up a new sample. It turns out it is possible to do inline diffs using <del>`<del>`</del> and <ins>`<ins>`</ins> (which GH does render) but they cannot be colorized and I found them really hard to spot in the output.

There's currently no limit; all changed styles are rendered. WRT dependents/independents:

* How do I get `Independents` and `Dependents` into my script? What should I `require`?
* Will they pick up new styles on disk, or do they only list items from whatever gem that hosts them?
* If an independent style changes, should I be checking all its dependents?

I will look at your point 2.

**Comment by retorquere** [comment]
Does anyone here know what `R099` means in `git --name-status`? I know `R` means renamed, but I don't know what the numbers mean; they're significant because a simple rename gave me `R100` and a rename + mod gave me `R099`, but I'd rather know rather than guess how to interpret these numbers.

**Comment by retorquere** [comment]
I've just taken (in)dependence from the path names.

**Comment by retorquere** [comment]
Currently set up so that if the tests fail, it will post the failure, if they pass, at most 10 styles are rendered, with independents taking precedence, and it will tell you how many there actually were.

**Comment by retorquere** [comment]
The current script lives on https://github.com/retorquere/styles/tree/reporting; the files of interest are:

* `.rspec` (adds file output for the rspec runner)
* `.travis.yml` (calls the script to post to github)
* `spec/post-to-github.rb`
* `spec/items.json` (the sample items)

**Comment by adam3smith** [comment]
> Does anyone here know what R099 means in git --name-status? 

from https://stackoverflow.com/a/35142442/1483360

> Status letters C and R are always followed by a score (denoting the percentage of similarity between the source and target of the move or copy). Status letter M may be followed by a score (denoting the percentage of dissimilarity) for file rewrites.

So makes perfect sense that a pure rename is 100% similar and a rename&mod is some number <100% similar.

**Comment by adam3smith** [comment]
I'm really happy with this as is -- we'd obviously want to add brief messages for the contributors to make sense of this, but in terms of the output I would just take this as in the last example. @rmzelle -- are you good to go with the current version?

I think just adding explanatory text during the merge process might be easiest so Rintze and I can debate the details there..

**Comment by rmzelle** [comment]
> I'm really happy with this as is

@adam3smith, I agree this looks great. With a few more changes it looks like we can retire Sheldon entirely, right? I think the only thing we're missing right now is the welcome comment after a PR is opened (and support for the "locales" repo).

@retorquere, with regard to the failed tests, it would be very nice if we could clean up the rspec output a little. I don't know if this can be configured directly within rspec, but the following lines should ideally be stripped from the output as they're not informative to the typical contributors of CSL styles:

>      Shared Example Group: "style" called from ./spec/styles_spec.rb:223
>      # ./spec/styles_spec.rb:169:in `block (2 levels) in '

The final block can also be deleted:

```
> Finished in 1 minute 28.2 seconds (files took 41.24 seconds to load)
> 156043 examples, 8 failures
> 
> Failed examples:
> 
> rspec ./spec/repository_spec.rb:21 # The CSL Style Repository may not contain any duplicate style titles
> rspec ./spec/styles_spec.rb[1388:6] # developing-world-bioethics: "template" link must point to an existing independent style
> rspec ./spec/styles_spec.rb[1625:15] # junk: style ID must be of the form "http://www.zotero.org/styles/" + style file name (without ".csl" extension, e.g. "http://www.zotero.org/styles/apa")
> rspec ./spec/styles_spec.rb[3567:4] # dependent/apa-cv: must be a dependent style (independent styles must be placed in the root directory)
> rspec ./spec/styles_spec.rb[3567:5] # dependent/apa-cv: "independent-parent" link must point to an existing independent style
> rspec ./spec/styles_spec.rb[3567:6] # dependent/apa-cv: may not have , , or  elements
> rspec ./spec/styles_spec.rb[3567:7] # dependent/apa-cv: may not have a "template" link
> rspec ./spec/styles_spec.rb[3567:9] # dependent/apa-cv: must have the same citation-format as its independent-parent
```

This would reduce https://github.com/retorquere/styles/pull/1#issuecomment-493754415 to:

<details><summary>Tests failed</summary><p>

    1) The CSL Style Repository may not contain any duplicate style titles
       Failure/Error: expect(TITLES.select { |_, styles| styles.length > 1 }).to eq({})
       
         expected: {}
              got: {"american psychological association 6th edition"=>["apa", "junk"]}
       
         (compared using ==)
       
         Diff:
         @@ -1 +1,2 @@
         +"american psychological association 6th edition" => ["apa", "junk"],
    
    2) developing-world-bioethics: "template" link must point to an existing independent style
       Failure/Error: expect(INDEPENDENTS_BASENAMES).to include(template_ID)
         expected ["vancouver-brackets-only-year-no-issue", "orthopedic-clinics-of-north-america", "asa-cssa-sssa", "ac...nschaften-und-landschaftsarchitektur", "revista-noesis", "haute-ecole-de-gestion-de-geneve-iso-690"] to include "apa-cv"
    
    3) junk: style ID must be of the form "http://www.zotero.org/styles/" + style file name (without ".csl" extension, e.g. "http://www.zotero.org/styles/apa")
       Failure/Error: expect(style.id).to eq("http://www.zotero.org/styles/#{basename}")
       
         expected: "http://www.zotero.org/styles/junk"
              got: "http://www.zotero.org/styles/apa"
       
         (compared using ==)
    
    4) dependent/apa-cv: must be a dependent style (independent styles must be placed in the root directory)
       Failure/Error: expect(style).to be_dependent
         expected `#.dependent?` to return true, got nil
    
    5) dependent/apa-cv: "independent-parent" link must point to an existing independent style
       Failure/Error: expect(parent_ID_link).to match(%r{^#{link_prefix}})
         expected nil to match /^http:\/\/www.zotero.org\/styles\//
    
    6) dependent/apa-cv: may not have , , or  elements
       Failure/Error: expect(style).not_to have_macro
         expected #has_macro? to return false, got true
    
    7) dependent/apa-cv: may not have a "template" link
       Failure/Error: expect(style).not_to have_template_link
         expected #has_template_link? to return false, got true
    
    8) dependent/apa-cv: must have the same citation-format as its independent-parent
       Failure/Error: parent = style.independent_parent_link[/[^\/]+$/]
       
       NoMethodError:
         undefined method `[]' for nil:NilClass
  </p></details>

Finally, a link to the build report would also be handy, and I see that some HTML/XML tags are currently not being escaped properly (`6) dependent/apa-cv: may not have , , or  elements`).

**Comment by retorquere** [comment]
The script can also post the welcome message, and I currently see a few ways to go about this; the script would scan the PR comments upon start

1. for a fixed template, and if not found, post it. This would mean the welcome gets repeated if the template changes.
2. for a given flair to be added to the welcome comment. It's sort of unlikely that the opening comment would have any flair. I think I can see flair in the API.
3. for a given user to have posted on the issue comments. This assumes the script would get a github token tied to a bot account, not a personal account. I would recommend this in any case, because it gets confusing if bots post under personal accounts.

As to trimming the output, I am looking at the rspec json logger to see if I can make sense of its output. It's possible to write a custom formatter to do whatever we want directly (such as generating markdown), but I could not find any documentation on this saying anything else than "it's possible".

**Comment by retorquere** [comment]
New sample of test failure log is up.

**Comment by rmzelle** [comment]
@retorquere, I think (3) makes the most sense. We currently use https://github.com/csl-bot to post the Sheldon comments (which is actually under control of Zotero), and we might be able to co-opt that account.

And where did you make the changes that resulted in https://github.com/retorquere/styles/pull/1#issuecomment-493895680? https://github.com/retorquere/styles/tree/reporting hasn't seen an update since 5/19.

Would you also be open to supporting the "locales" repo? We use Sheldon in very similar capacity there (just with separate templates: https://github.com/citation-style-language/Sheldon/tree/master/templates). We probably wouldn't need any citation rendering there, but it would e.g. be nice to have the GitHub comments with test failure, and it would allow us to retire Sheldon altogether.

**Comment by retorquere** [comment]
I forgot to push the latest changes, they're now up. Yeah sure, I haven't looked at the locales repo, but if it substantially does the same thing, that seems sensible. I'll have a look.

I'll add the scanning code for point 3.

**Comment by retorquere** [comment]
Turns out it's not too hard to make a github-installable gem to share the code. Would you guys prefer the test-assets (like `items.json`) to live with the gem or in the respective repos (`styles`/`locales`)?

**Comment by adam3smith** [comment]
> Turns out it's not too hard to make a github-installable gem to share the code. Would you guys prefer the test-assets (like items.json) to live with the gem or in the respective repos (styles/locales)?

I don't have a strong opinion but my first reaction would be to leave them in the respective repos because Rintze and I have those cloned locally already, so easier to update. But if there's any reason to keep them in the gem (speed, some other best practice reason), this is a minor consideration and I'd be happy with that, too.

**Comment by retorquere** [comment]
There's no real benefit, but I saw that Sheldon also had local assets for styles and locales, and perhaps the `items.json` could be shared.

**Comment by rmzelle** [comment]
> There's no real benefit, but I saw that Sheldon also had local assets for styles and locales, and perhaps the `items.json` could be shared.

I'm not sure we need items.json for the locales repo, as the benefit of citation rendering is rather limited there.

**Comment by retorquere** [comment]
Alright, local it is.

**Comment by retorquere** [comment]
There's the other issue though that the sheldon gem now relies on files being in specific places in another repo. I don't mind, personally.

What did/should sheldon report/do for the locales repo?

**Comment by rmzelle** [comment]
> There's the other issue though that the sheldon gem now relies on files being in specific places in another repo. I don't mind, personally.

I don't think this really matters for us as long as there is some documentation somewhere on what file-dependencies there are.

> What did/should sheldon report/do for the locales repo?

See the three comments by @csl-bot in e.g. https://github.com/citation-style-language/locales/pull/188. We'd have a welcome comment, and for each commit a success or failure comment. We'd want to insert the test failures directly again here as well, instead of just linking to the Travis CI report.

It would be good if your implementation would still provide links to the Travis CI reports, by the way, for both "styles" and "locales" repositories (e.g. underneath the test with a "(see <here> for the full Travis CI test report)".

**Comment by retorquere** [comment]
Gemification done:

* [Sheldon](https://github.com/retorquere/Sheldon/tree/travis)
* [styles](https://github.com/retorquere/locales/tree/reporting)
* [locales](https://github.com/retorquere/locales/tree/reporting)

**Comment by retorquere** [comment]
I've added a README to the Sheldon branch that explains the expectations it has on the styles and locales repos; all files live in `spec/sheldon` in those respective repos currently, and the messages Sheldon will post can be edited there.

**Comment by retorquere** [comment]
I'm terribly sorry about this, but I had forgotten that in PRs, secret variables like github tokens are not set (for good reasons). This means PR-builds cannot post directly to GH. I've been talking to Travis support about this (who are super responsive!), and it looks like the best option is to mostly keep Sheldon but also keep the stuff I've done and have them cooperate.

The reason now-Sheldon can't (reasonably) do everything on its own is that you really need a checked out repo at the exact point where the PR runs as well as access to the master branch and have the travis vars available -- we'd be trying to replicate the Travis build environment on Heroku (which is where shel-bot runs right?)

The coop solution would be to keep shel-gem mostly as-is, but have it not post to github but either upload its output to transfer.sh or just output it to the console. shel-bot would then parse out either the transfer.sh response url or the shel-gem output from the travis log, and post that.

There's upsides and downsides to each of these coop modes:

* transfer.sh would look cleanest (it's just a simple URL), but it puts a service between the two sides of sheldon, which may be fragile.
* Outputting the comments would be most robust, but they would be user-visible, and not in a format that would necessarily be nice to look at (a mixture of raw HTML and markdown).

I don't currently see a way around this. I'm trying to think of a format which would look good enough to post in the log that would also be re-parsable for shel-bot.

**Comment by adam3smith** [comment]
I think it makes sense to go with the transfer.sh option and see if we're having issues with stability. Given that one of the reasons for doing this is to improve the experience for style contributors, clean output would be important.

**Comment by retorquere** [comment]
I'm trying a few services - transfer.sh is really slow (10-20 seconds) and that has me worried about the order of execution of shel-bot and shel-gem. 0x0.st is substantially faster (tenth of a second) but it's proven hard to get going with ruby and the project owner is doing the best impression of an elitist prick I've seen in a long while when I asked for support. File.io is also substantially faster (same as 0x0.st) but files can only be requested once there and I'm worried that will complicate debugging.

**Comment by retorquere** [comment]
0x0.st explicitly forbids "spamming the service with CI build assets" and file.io has both a monthly limit of 100 uploads (and it's probably enforced by IP address so from a CI service we'd be pooled in with anyone else using it) and there's some kind of rate limit going on (I sometimes get "too many requests" even from my home system).

If anyone knows other services of the kind, I'd love to hear about it, but I'm currently looking for different solutions altogether.

**Comment by retorquere** [comment]
One other option would be to set up an S3 or backblaze bucket which allows anon uploads but only authenticated downloads, with an auto-expiry policy to keep the bucket size down. I use this setup for BBT logs and I've only once broken the 8ct/month limit which was when I was furiously putting out new builds in the Z5 port. Benefit of this would be that there'd need to be nothing in the logs -- shel-bot could just pick it up at an agreed name.

**Comment by retorquere** [comment]
There's one other way that just struck me --  technically possible but ho-hum from a beauty point of view. The travis log renderer honors ansi escape sequences, so I could output the shel-bot message, erase it using ansi backspaces, which would remove it from view but it would remain in the actual log where I could parse it out.

**Comment by retorquere** [comment]
There's an ungodly amount of magic going on in rspec -- I'm getting 

```
undefined method `assert_equal' for nil:NilClass
```

when I run `rspec` on shel-bot. I've so far traced it back to the minitest context (`ctx`) being nil in `minitest-5.11.3/lib/minitest/spec.rb`. Anyone here experience with rspec?

**Comment by retorquere** [comment]
In a clean rvm setup, I'm hitting

```
$ rspec
/Users/emile/.rvm/rubies/ruby-2.4.1/lib/ruby/site_ruby/2.4.0/rubygems.rb:283:in `find_spec_for_exe': can't find gem rspec-core (>= 0.a) with executable rspec (Gem::GemNotFoundException)
	from /Users/emile/.rvm/rubies/ruby-2.4.1/lib/ruby/site_ruby/2.4.0/rubygems.rb:302:in `activate_bin_path'
	from /usr/local/lib/ruby/gems/2.6.0/bin/rspec:23:in `<main>'
```

**Comment by retorquere** [comment]
Wait, I don't need rspec -- I just need to run `rake`. But when I run `rake`, I get

```
Fabulous run in 0.000468s, 0.0000 runs/s, 0.0000 assertions/s.

0 runs, 0 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by retorquere** [comment]
This is driving me nuts. What I see in the Rakefile seems to use minitest, but then the actual spec files use rspec's `describe`/`it` syntax. @inukshuk, can you give me a push in the right direction?

**Comment by retorquere** [comment]
Alright, more clarity; the csl gems pull in newer versions of rspec it seems, and the existing test setup does not like this. No errors, but no tests ran.

The quick way to solve this is to keep shel-bot and shel-gem on different branches. I do worry a little about the technological debt shel-bot is going to accumulate over time. I had tried to merge shel-bot and shel-gem (which should be possible), but realistically, that would require some gems in shel-bot to be updated, and due to the amount of magic going on in rspec, it's not easy to find out which gem is injecting what into the various namespaces.

**Comment by inukshuk** [comment]
@retorquere I've not been following along, but on a quick glance `Sheldon` does not use `rspec` but rather `minitest`. The csl/citeproc gems use rspec, but only as dev dependencies -- those should not be pulled in when you install them (unless you specifically use dev dependencies).

**Comment by retorquere** [comment]
My bad.

I've been able to get it to work by [pinning gems to the exact versions I get when the tests run on Travis](https://github.com/retorquere/Sheldon/blob/gem/sheldon.gemspec#L33). I've not been able to get the tests to run bu just having the gems all at their latest versions -- `rake` runs without errors but no tests are actually ran.

If this pinning is OK, I can proceed to merge the shel-gem part; it doesn't actually touch shel-bot (yet), just adds the gem-installable executable to the existing Sheldon so that the travis counterpart can be easily installed on `styles` and `locales`.

**Comment by inukshuk** [comment]
I'm a bit confused by that gemspec file. Most gems don't package test/spec files at all, in my experience; the test files also don't seem to be added to `spec.files` here, so maybe that's why no tests ran? The file also pulls in rspec even though it's not being used, that doesn't seem right. In any case, pinning versions of dev and test dependencies is much easier to do using bundler / Gemfile.lock -- I'd really try to use that instead of specifying everything as gem dependencies. Specifying something like `puma` is also potentially restricting since you could use other rack-compatible servers to run this. Furthermore, the `bin` folder` only contained bundler binstubs, which are just there for convenience really, it's not something that should be included in a gem. 

That said, I'm really out of the loop here, so I may be missing the point.

**Comment by retorquere** [comment]
Ah no, the gemspec file isn't complete yet, but for the shel-bot part nothing changes; the gemspec is really only being used by gem consumers. I can run `rake` in the current state and the tests run, it's just when I upgraded the gems that the tests don't run. I don't intend to add the spec files to `spec.files`.

Using the gemspec is really the same as using the Gemfile from the point of view of shel-bot -- a lockfile is also generated as normal. That said, I'm no expert on gem packaging, I'm just cribbing things together best I can.

Turns out `puma` isn't required, I just at some point added everything that I saw was being pulled in previously to just get the tests going. The tests run without `puma` being brought in just fine, so I've removed it. I'm going through the pinned gems one by one to see which I can remove from pinning.

The `bin` directory also include the executable I intend to run on Travis, so it's no longer just the stubs.

In the end, the idea is that the gem-part offers a ruby script that generates a GH-friendly summary of the test results, which the bot-part can pick up. I could just have either the gem-part on a separate branch, or even a separate repo, but given that these two would work so closely together, at least the same repo would make most sense to me.

**Comment by inukshuk** [comment]
Is the part that's going to be run on Travis also run by the stateless web service on Heroku?

**Comment by retorquere** [comment]
No, but the part that's going to be ran on heroku will expect there to be something in the travis log that's put there by the script in `bin`.

**Comment by retorquere** [comment]
I'm going to start moving stuff back into the Gemfile -- I had misinterpreted the gemspec stuff I found to say that everything needs to be there, which is not the case.

**Comment by inukshuk** [comment]
Cool, I think that's better (I'd just leave runtime deps in the gemspec). If the Travis script does not run on Heroku, I'd also put its dependencies into a separate Group in the Gemfile which is not installed on Heroku. But in general, maybe it would be easier to use a separate repo for this (because I'm not sure that you could actually keep dependencies defined in the gemspec file out of the bundle created on Heroku).

**Comment by retorquere** [comment]
It may be possible to have the gem deps not installed on heroku; the current gemfile adds a section called `travis` (https://github.com/retorquere/Sheldon/blob/gem/Gemfile#L29), but I don't know how heroku chooses what sections to install.

**Comment by retorquere** [comment]
(other than that it turns out I only had to pin hashdiff because newer versions of hashdiff have a namespace conflict with another gem).

**Comment by retorquere** [comment]
OK, I think I have things prepped now. Shel-bot should be unaffected other for picking up the build report from the log, and the sheldon gem now lives in a single script. I'm going to deploy it to my own heroku instance to see how it works out.

**Comment by retorquere** [comment]
Two questions:

1. How can I verify that my copy of shel-bot is receiving events and acting on them? I have my copy set up on https://retorquere-sheldon.herokuapp.com/ but I'm not seeing any responses on my [own PR](https://github.com/retorquere/styles/pull/1).
2. Perhaps related: it looks like PRs submitted by the repo owner do not get treated as PRs; @adam3smith, could you merge the reporting branch again and push to see if that's the difference?

**Comment by rmzelle** [comment]
>     1. How can I verify that my copy of shel-bot is receiving events and acting on them? I have my copy set up on https://retorquere-sheldon.herokuapp.com/ but I'm not seeing any responses on my [own PR](https://github.com/retorquere/styles/pull/1).

Maybe you're missing a webhook (https://github.com/citation-style-language/styles/blob/0e4b1f80ed20ccf3a88febebf2c526aa734d80e1/.travis.yml#L14)?

**Comment by retorquere** [comment]
I have it figured out (and tested), the PRs are ready.

**Comment by adam3smith** [comment]
> I have it figured out (and tested), the PRs are ready.

... and waiting for your input: https://github.com/citation-style-language/Sheldon/pull/15#issuecomment-499906106

**Comment by retorquere** [comment]
OK, I think everything is in place now. Sheldon is running on heroku-18, and `styles`/`locales` now automatically fetch the latest sheldon reporter during the tests, so maintenance can be done on this repo for the entire setup. I think that is that then?

**Comment by rmzelle** [comment]
Thanks both for getting this in place!

**Comment by retorquere** [comment]
All for a good cause https://imgur.com/a/gruF37o
